hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotCache.java

```diff
@@ -224,29 +225,33 @@ public ReferenceCounted<IOmMetadataReader, SnapshotCache> get(String key,
    * @param key snapshot table key
    */
   public void release(String key) {
-    ReferenceCounted<IOmMetadataReader, SnapshotCache> rcOmSnapshot =
-        dbMap.get(key);
-    Preconditions.checkNotNull(rcOmSnapshot,
-        "Key '" + key + "' does not exist in cache");
-
-    if (rcOmSnapshot.decrementRefCount() == 0L) {
-      synchronized (pendingEvictionList) {
-        // Eligible to be closed, add it to the list.
-        pendingEvictionList.add(rcOmSnapshot);
-        // The cache size might have already exceed the soft limit
-        // Thus triggering cleanup() to check and evict if applicable
-        cleanup();
+    dbMap.compute(key, (k, v) -> {
+      if (v == null) {
+        throw new IllegalArgumentException(
+            "Key '" + key + "' does not exist in cache");
       }
-    }
+
+      if (v.decrementRefCount() == 0L) {
+        synchronized (pendingEvictionList) {
+          // Eligible to be closed, add it to the list.
+          pendingEvictionList.add(v);
+          // The cache size might have already exceed the soft limit
+          // Thus triggering cleanup() to check and evict if applicable
+          cleanup();
+        }
+      }
+
+      return v;
```

Comment:
Return v here will put it back in the dbMap, but if it gets removed in cleanup() above, I don' t think that is what we want.

Reply:
Good point. Replaced this line with `return dbMap.get(k);`

Might look hacky, but it works. Tested with a snippet:

```java
import java.util.concurrent.ConcurrentHashMap;

class Untitled {
  public static void main(String[] args) {
    ConcurrentHashMap map = new ConcurrentHashMap<String, String>();
    String k1 = "k1";
    map.put(k1, "val1");
    map.compute(k1, (k, v) -> {
      System.out.format("k=%s, v=%s\n", k, v);
      map.remove(k);
      return map.get(k);
    });
    System.out.format("k=%s, v=%s\n", k1, map.get(k1));
  }
}
```

Output:
```
k=k1, v=val1
k=k1, v=null
```

Reply:
nvm. turns out it is a bad idea to update other mappings inside a `map.compute()` anyways according to its javadoc. In this case it corrupts the map size count, and it caught by my UT:

https://github.com/apache/ozone/actions/runs/5396424029/jobs/9800041580?pr=4567#step:6:2594

Repro snippet:

```java
import java.util.concurrent.ConcurrentHashMap;

class Untitled {
  public static void main(String[] args) {
    ConcurrentHashMap map = new ConcurrentHashMap<String, String>();
    String k1 = "k1";
    map.put(k1, "v1");
    map.put("k2", "v2");
    System.out.println("- Map dump 1");
    map.forEach((k, v) -> {
      System.out.format("k=%s, v=%s\n", k, v, map.size());
    });
    System.out.println();
    
    System.out.println("- compute()");
    map.compute(k1, (k, v) -> {
      System.out.format("k=%s, v=%s, size=%d\n", k, v, map.size());
      map.remove(k);
      System.out.format("k=%s, v=%s, size=%d\n", k, v, map.size());
      return map.get(k);
    });
    System.out.println();
    // Map size becomes incorrect at this point
    System.out.format("k=%s, v=%s, size=%d\n", k1, map.get(k1), map.size());
    
    System.out.println("- Map dump 2");
    map.forEach((k, v) -> {
      System.out.format("k=%s, v=%s\n", k, v, map.size());
    });
  }
}
```

Output:
```
- Map dump 1
k=k1, v=v1
k=k2, v=v2

- compute()
k=k1, v=v1, size=2
k=k1, v=v1, size=1

k=k1, v=null, size=0
- Map dump 2
k=k2, v=v2
```

You can see map's `size=0` while it obviously still has `k2` entry in it.

I have moved `cleanup()` out of `compute()` and added additional syncing.

Reply:
Since the ref counting is still done inside dbMap's monitor lock, the previously mentioned issue [here](https://github.com/apache/ozone/pull/4567#discussion_r1232881842) and [here](https://github.com/apache/ozone/pull/4567#discussion_r1232883122) should not revive.

And `cleanup()` is now wrapped inside `synchronized`.